ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1,165 +1,408 @@
+// params can be either BSDFIsotropicParams or BSDFAnisotropicParams
+Spectrum irr_bsdf_cos_eval(in vec3 L, in irr_glsl_IsotropicViewSurfaceInteraction inter, in mat2 dUV)
 {
-	bsdf::SBSDFUnion retval;
-	uint32_t bitfields = _node->type;
-	switch (_node->type)
+	instr_stream_t eval_instrStream = getEvalStream();
+	
+	return runEvalStream(eval_instrStream, L);
+}
+#endif
+
+#ifndef _IRR_COMPUTE_LIGHTING_DEFINED_
+#define _IRR_COMPUTE_LIGHTING_DEFINED_
+vec3 irr_computeLighting(inout irr_glsl_IsotropicViewSurfaceInteraction out_interaction, in mat2 dUV)
+{
+	vec3 emissive = irr_glsl_decodeRGB19E7(InstData.data[InstanceIndex].emissive);
+
+	vec3 campos = irr_glsl_MC_getCamPos();
+	//irr_glsl_BSDFIsotropicParams params;
+	//params.L = campos-WorldPos;
+	out_interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, normalize(Normal));
+
+	return irr_bsdf_cos_eval(normalize(campos-WorldPos), out_interaction, dUV)*1000.0/dot(params.L,params.L) + emissive;
+}
+#endif
+
+void main()
+{
+	mat2 dUV = mat2(dFdx(UV),dFdy(UV));
+
+	InstanceData instData = InstData.data[InstanceIndex];
+#ifdef TEX_PREFETCH_STREAM
+	runTexPrefetchStream(getTexPrefetchStream(), dUV);
+#endif
+#ifdef NORM_PRECOMP_STREAM
+	runNormalPrecompStream(getNormalPrecompStream(), dUV);
+#endif
+
+	irr_glsl_IsotropicViewSurfaceInteraction inter;
+	vec3 color = irr_computeLighting(inter, dUV);
+
+	OutColor = vec4(color, 1.0);
+}
+)";
+
+_IRR_STATIC_INLINE_CONSTEXPR const char* PIPELINE_LAYOUT_CACHE_KEY = "irr/builtin/pipeline_layout/loaders/mitsuba_xml/default";
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERTEX_SHADER_CACHE_KEY = "irr/builtin/specialized_shader/loaders/mitsuba_xml/default";
+
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t PAGE_TAB_TEX_BINDING = 0u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t PHYS_PAGE_VIEWS_BINDING = 1u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t PRECOMPUTED_VT_DATA_BINDING = 2u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t INSTR_BUF_BINDING = 3u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t BSDF_BUF_BINDING = 4u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t INSTANCE_DATA_BINDING = 5u;
+_IRR_STATIC_INLINE_CONSTEXPR uint32_t DS0_BINDING_COUNT_WO_VT = 4u;
+
+template <typename AssetT>
+static void insertAssetIntoCache(core::smart_refctd_ptr<AssetT>& asset, const char* path, IAssetManager* _assetMgr)
+{
+	asset::SAssetBundle bundle({ asset });
+	_assetMgr->changeAssetKey(bundle, path);
+	_assetMgr->insertAssetIntoCache(bundle);
+}
+template<typename AssetType, IAsset::E_TYPE assetType>
+static core::smart_refctd_ptr<AssetType> getBuiltinAsset(const char* _key, IAssetManager* _assetMgr)
+{
+	size_t storageSz = 1ull;
+	asset::SAssetBundle bundle;
+	const IAsset::E_TYPE types[]{ assetType, static_cast<IAsset::E_TYPE>(0u) };
+
+	_assetMgr->findAssets(storageSz, &bundle, _key, types);
+	if (bundle.isEmpty())
+		return nullptr;
+	auto assets = bundle.getContents();
+	//assert(!assets.empty());
+
+	return core::smart_refctd_ptr_static_cast<AssetType>(assets.begin()[0]);
+}
```

Comment:
dont we already have the code for this somewhere else?

Reply:
possible, in some other loader

Reply:
@AnastaZIuk ever seen anything similar in other places in our codebase?

Reply:
@AnastaZIuk did you do any changes?

Reply:
i think it would be good idea to just put this function into this loader base class @AnastaZIuk is supposed to make

Reply:
so the base was made?